# Lab04 - pandas II
#### Name: Matt McLaughlin 
#### Class: CSCI 349 - Intro to Data Mining 
#### Semester: 2020SP 
#### Instructor: Brian King

In [1]:
import sys
import numpy as np
import pandas as pd

In [2]:
days = ["Mon","Tue","Wed","Thurs","Fri","Student"]
scores = pd.DataFrame([pd.Series([9.5, 8.75, 8, 10, 7.75, "Bob"], index=days, name="week_1"),
pd.Series([9, 8, 10, 8.75, 7.25, "Jane"], index=days, name="week_1"),
pd.Series([8.5, 0, 9.75, 9, 6, "Bob"], index=days, name="week_2"),
pd.Series([7, 8.25, 9.25, 0, 8, "Jane"], index=days, name="week_2"),
pd.Series([7, 8.5, 9.25, 0, 0, "Bob"], index=days, name="week_3"),
pd.Series([8.5, 8.25, 9, 8, 7.5, "Jane"], index=days, name="week_3"),
pd.Series([9, 7, 8.5, 8, 8, "Bob"], index=days, name="week_4"),
pd.Series([9.25, 9.25, 8.5, 7.75, 7.5, "Jane"], index=days,name="week_4") ])
scores

Mon   Tue    Wed  Thurs   Fri Student
week_1  9.50  8.75   8.00  10.00  7.75     Bob
week_1  9.00  8.00  10.00   8.75  7.25    Jane
week_2  8.50  0.00   9.75   9.00  6.00     Bob
week_2  7.00  8.25   9.25   0.00  8.00    Jane
week_3  7.00  8.50   9.25   0.00  0.00     Bob
week_3  8.50  8.25   9.00   8.00  7.50    Jane
week_4  9.00  7.00   8.50   8.00  8.00     Bob
week_4  9.25  9.25   8.50   7.75  7.50    Jane

### 1) [P] Report the shape, the number of dimensions, the size of the data, and the data types of each column. Print each separately. Then create an additional code cell that shows the results of scores.info().

In [3]:
print("Shape: " + str(scores.shape), '\n')
print("Number of dimensions: " + str(scores.ndim), '\n')
print("Size of data (bytes): " + str(scores.size * scores.dtypes.itemsize), '\n') #This doesn't seem right
print("Data types: " + str(scores.dtypes))

Shape: (8, 6) 

Number of dimensions: 2 

Size of data (bytes): 384 

Data types: Mon        float64
Tue        float64
Wed        float64
Thurs      float64
Fri        float64
Student     object
dtype: object


C:\Users\Matt\Anaconda3\envs\csci349\lib\site-packages\ipykernel_launcher.py:3: FutureWarning: Series.itemsize is deprecated and will be removed in a future version
  This is separate from the ipykernel package so we can avoid doing imports until


In [4]:
scores.info()

<class 'pandas.core.frame.DataFrame'>
Index: 8 entries, week_1 to week_4
Data columns (total 6 columns):
Mon        8 non-null float64
Tue        8 non-null float64
Wed        8 non-null float64
Thurs      8 non-null float64
Fri        8 non-null float64
Student    8 non-null object
dtypes: float64(5), object(1)
memory usage: 448.0+ bytes


### 2) [P] Show two different ways to report the number of observations in scores.

In [5]:
#Technique 1
print(len(scores.index), '\n')

#Technique 2
print(scores.shape[0])

8 

8


### 3) [P] Show the variables (i.e. columns) in scores

In [6]:
scores.columns

Index(['Mon', 'Tue', 'Wed', 'Thurs', 'Fri', 'Student'], dtype='object')

### 4) [P] There are repetitive names in the index. This happens quite a bit, depending on how the rows are to be named. For now. print the unique index names in scores

In [7]:
scores.index.unique()

Index(['week_1', 'week_2', 'week_3', 'week_4'], dtype='object')

### 5) [P] Rename the 'Thurs' column header to be 'Thu' . (Hint: Changing scores.columns[3] directly will not work, not to mention it would be a poor approach to hard code a location this way! Magic numbers be bad when referencing index or column locations. They may change!) Show the new scores data frame. This should be the same, except with Thurs changed to Thu.

In [8]:
scores = scores.rename(columns = {"Thurs":"Thu"})
#Show new scores
scores

Mon   Tue    Wed    Thu   Fri Student
week_1  9.50  8.75   8.00  10.00  7.75     Bob
week_1  9.00  8.00  10.00   8.75  7.25    Jane
week_2  8.50  0.00   9.75   9.00  6.00     Bob
week_2  7.00  8.25   9.25   0.00  8.00    Jane
week_3  7.00  8.50   9.25   0.00  0.00     Bob
week_3  8.50  8.25   9.00   8.00  7.50    Jane
week_4  9.00  7.00   8.50   8.00  8.00     Bob
week_4  9.25  9.25   8.50   7.75  7.50    Jane

### 6) [P] Now, make a bigger change. Let's rename 'week_1' to be 'w1', 'week_2' to be 'w2', and so on. Also, suppose you decide to use only 2 letter abbreviations for the days. Rename the days in the column names to be 'Mo', 'Tu', 'We', 'Th', 'Fr'.

In [9]:
scores = scores.rename(index={"week_1":"w1","week_2":"w2","week_3":"w3","week_4":"w4"}, columns = {"Mon":"Mo", "Tue":"Tu", "Wed":"We","Thu":"Th", "Fri":"Fr"})
scores

Mo    Tu     We     Th    Fr Student
w1  9.50  8.75   8.00  10.00  7.75     Bob
w1  9.00  8.00  10.00   8.75  7.25    Jane
w2  8.50  0.00   9.75   9.00  6.00     Bob
w2  7.00  8.25   9.25   0.00  8.00    Jane
w3  7.00  8.50   9.25   0.00  0.00     Bob
w3  8.50  8.25   9.00   8.00  7.50    Jane
w4  9.00  7.00   8.50   8.00  8.00     Bob
w4  9.25  9.25   8.50   7.75  7.50    Jane

### 7) [P] Compare the type of the expression scores['Mo'] vs. scores[['Mo']]. What is the difference? (HINT: use the type() Python function on both.)

In [10]:
print(type(scores["Mo"]))
print(type(scores[['Mo']]))

<class 'pandas.core.series.Series'>
<class 'pandas.core.frame.DataFrame'>


#### A single bracket returns a Series, the column in the DataFrame. Two brackets returns a DataFrame with only 1 column

### Read about the describe() method of data frames. (FYI - This is essentially the equivalent to the summary() function in R.)

### 8) [P] Demonstrate the describe() method on scores. What type does it return? From your observations of the output, which day had the largest standard deviation in quiz scores? (You can just hard code the answer to this question. You'll write code to do this next.)

In [11]:
scores.describe()

Mo      Tu         We         Th        Fr
count  8.000000  8.0000   8.000000   8.000000  8.000000
mean   8.468750  7.2500   9.031250   6.437500  6.500000
std    0.967669  3.0000   0.673974   4.037215  2.702512
min    7.000000  0.0000   8.000000   0.000000  0.000000
25%    8.125000  7.7500   8.500000   5.812500  6.937500
50%    8.750000  8.2500   9.125000   8.000000  7.500000
75%    9.062500  8.5625   9.375000   8.812500  7.812500
max    9.500000  9.2500  10.000000  10.000000  8.000000

#### It returns a DataFrame with a variety of statistics about each column in the DataFrame. Thursday has the largest standard deviation.

### 9) [P] While you could look at the output of describe() and infer information, it's more important to write code to output the specific results you are looking for. Write the Python code that shows the largest standard deviation and the day it occurred without using describe().

In [12]:
print("Max STD: " + str(scores.std().max()) + " \non: " + str( (scores.std() == scores.std().max()).idxmax(axis=1) ))

Max STD: 4.037215270809465 
on: Th


### OK. Take a moment and look over that output. If you've been doing everything correctly, then that result should be ~4.0372 occurring on Thursday, and that is way too large! Think back to your stats course. What does standard deviation mean? It's the deviation from the mean! That's a large spread in the data! Why? There are two missing values represented by 0 that are being used to compute the deviation, and that's not what we want. Missing data are usually not included with summary statistics (and many other calculations!) We need to deal with missing data properly. So, instead, when we have missing values in data, we almost always use a na.nan value in its place to represent that the data is missing in that specific cell.

### 10) [P] Write the code that changes all 0.0 entries to np.nan. (NOTE: You should be able to do this with just one line of code using pandas selection techniques!) Then, show scores. All 0.0 entries should be replaced with NaN in the output.

In [13]:
scores = scores.replace(0, np.nan)
scores

Mo    Tu     We     Th    Fr Student
w1  9.50  8.75   8.00  10.00  7.75     Bob
w1  9.00  8.00  10.00   8.75  7.25    Jane
w2  8.50   NaN   9.75   9.00  6.00     Bob
w2  7.00  8.25   9.25    NaN  8.00    Jane
w3  7.00  8.50   9.25    NaN   NaN     Bob
w3  8.50  8.25   9.00   8.00  7.50    Jane
w4  9.00  7.00   8.50   8.00  8.00     Bob
w4  9.25  9.25   8.50   7.75  7.50    Jane

### At this point, now you need to start paying attention to NaN values in your data. These represent missing values, and are important to ignore in most calculations, especially preliminary EDA. We will learn about many techniques for dealing with missing values soon. Many functions will have an optional parameter such as skipna to make sure that whatever you are trying to do does not include these missing data. Pay attention to those parameters. Most of the time, the default parameter values for handling NaN values is to skip or ignore them.

### 11) [P] Show the output of describe() again. Yes, this is better! Note how much more informative this output is when properly dealing with missing data. Copy your code from above to output the day with the largest standard deviation in quiz scores. You should have a substantially different result than you did before!

In [14]:
scores.describe()

Mo        Tu         We         Th        Fr
count  8.000000  7.000000   8.000000   6.000000  7.000000
mean   8.468750  8.285714   9.031250   8.583333  7.428571
std    0.967669  0.698638   0.673974   0.846562  0.687906
min    7.000000  7.000000   8.000000   7.750000  6.000000
25%    8.125000  8.125000   8.500000   8.000000  7.375000
50%    8.750000  8.250000   9.125000   8.375000  7.500000
75%    9.062500  8.625000   9.375000   8.937500  7.875000
max    9.500000  9.250000  10.000000  10.000000  8.000000

In [15]:
print("Max STD: " + str(scores.std().max()) + " \non: " + str( (scores.std() == scores.std().max()).idxmax(axis=1) ))

Max STD: 0.9676693281133947 
on: Mo


### 12) [P] Show the output of scores.values. What does the .values attribute do?

In [16]:
scores.values

array([[9.5, 8.75, 8.0, 10.0, 7.75, 'Bob'],
       [9.0, 8.0, 10.0, 8.75, 7.25, 'Jane'],
       [8.5, nan, 9.75, 9.0, 6.0, 'Bob'],
       [7.0, 8.25, 9.25, nan, 8.0, 'Jane'],
       [7.0, 8.5, 9.25, nan, nan, 'Bob'],
       [8.5, 8.25, 9.0, 8.0, 7.5, 'Jane'],
       [9.0, 7.0, 8.5, 8.0, 8.0, 'Bob'],
       [9.25, 9.25, 8.5, 7.75, 7.5, 'Jane']], dtype=object)

#### .values returns a numpy representation of the DataFrame

### 13) [P] What is the mean quiz score for each day? Do not use describe()

In [17]:
#I'm not sure if this is asking for each day Mo-Tu-We...   OR   Mo w1 - Tu w1 - We w1...
#Solution for option 1
scores.mean(0)

Mo    8.468750
Tu    8.285714
We    9.031250
Th    8.583333
Fr    7.428571
dtype: float64

In [18]:
#Solution for option 2
scores.groupby(by=scores.index).mean()

Mo     Tu     We     Th    Fr
w1  9.250  8.375  9.000  9.375  7.50
w2  7.750  8.250  9.500  9.000  7.00
w3  7.750  8.375  9.125  8.000  7.50
w4  9.125  8.125  8.500  7.875  7.75

### 14) [P] What is the mean quiz score over the entire dataset? (HINT: there are many ways to do this, and none should involve writing any loops! The only right answer is 8.375. If you aren't getting that, then you are not handling NaN values correctly. You only want the mean over all scores that are NOT NaN!)

In [19]:
scores.iloc[:,0:5].stack().mean()

8.375

### 15) [P] Now, show the mean for each week over all students. Again, you are not counting missing quizzes into the average. (Yes, technically you could write a loop to compute this exhaustively. However, you should rarely need to do that. Use pandas!) You have the correct result if you report the following (shown with 5 places of significant digits):

In [20]:
scores.mean(1).groupby(by=scores.index).mean()
# At first I did
#scores.groupby(by=scores.index).mean().mean(1)
# but it gave slightly incorrect answers. Interesting.

w1    8.70000
w2    8.21875
w3    8.25000
w4    8.27500
dtype: float64

### 16) [P] Select the data frame for only Monday and Student variables using .loc, and again using .iloc

In [21]:
scores.loc[:,["Mo", "Student"]]

Mo Student
w1  9.50     Bob
w1  9.00    Jane
w2  8.50     Bob
w2  7.00    Jane
w3  7.00     Bob
w3  8.50    Jane
w4  9.00     Bob
w4  9.25    Jane

In [22]:
scores.iloc[:,[0,5]]

Mo Student
w1  9.50     Bob
w1  9.00    Jane
w2  8.50     Bob
w2  7.00    Jane
w3  7.00     Bob
w3  8.50    Jane
w4  9.00     Bob
w4  9.25    Jane

### 17) [P] Show a new DataFrame containing the week as an index (like the original), but the mean of their scores and the count of the quizzes they took. 

In [23]:
#https://www.geeksforgeeks.org/adding-new-column-to-existing-dataframe-in-pandas/
df = pd.DataFrame(scores.loc[:,"Student"])

studentMeans = scores.mean(1)
studentCounts = scores.iloc[:,0:5].notna().sum(1)

df['count'] = studentCounts
df['mean'] = studentMeans
df

Student  count    mean
w1     Bob      5  8.8000
w1    Jane      5  8.6000
w2     Bob      4  8.3125
w2    Jane      4  8.1250
w3     Bob      3  8.2500
w3    Jane      5  8.2500
w4     Bob      5  8.1000
w4    Jane      5  8.4500

### 18) [P] Show the number of times Friday's score was greater than 7.25 for each student. Start by selecting your data with the [ ] operator.

In [24]:
#Success 1 - Neatest
fri = scores.loc[:,["Fr","Student"]]
fri["Fr"] = (fri.loc[:,"Fr"] > 7.25)
fri.groupby(by="Student").sum()

#Success 2
#boolArr = (scores.iloc[:,0:5] > 7.25) #Convert scores to T/F
#boolArr["Student"] = scores.iloc[:,5] #Reconnect Student column - Fugly but necessary
#boolArr.loc[:,["Fr","Student"]].groupby(by="Student").sum() #Trim to Fr and Student columns, count # True

#Success 3
#Use ICE
#bob = scores.loc[scores["Student"] == "Bob"]["Fr"]
#jane = scores.loc[scores["Student"] == "Jane"]["Fr"]
#pd.Series({"Jane":(jane>7.25).sum(), "Bob":(bob>7.25).sum()})

Fr
Student     
Bob      2.0
Jane     3.0

### 19) [P] Use the where() method on scores to repeat the previous exercise. (HINT: You might need to use dropna()). You should have the same output

In [25]:
fri19 = scores.loc[:,["Fr","Student"]]
fri19.where(fri19["Fr"]>7.25).groupby(by="Student").count()

Fr
Student    
Bob       2
Jane      3

### 20) [P] Select the scores that were greater than the mean score over the entire dataset (without the missing values). Again, no hard coding values here (i.e. do NOT hard code the mean, but recompute it. Then, report the number of scores for each day that exceeded that global mean. Your result should look like: This suggests that Wednesday seems to have the best quiz performance, whereas Friday... not so much.

In [26]:
days_only = scores.iloc[:,0:5]
total_mean = scores.iloc[:,0:5].stack().mean()
days_only.where(days_only > total_mean).count()

#This isn't what you wanted but I like it better
#days_only.where(days_only > total_mean).groupby(by=scores.index).count()

Mo    6
Tu    3
We    7
Th    3
Fr    0
dtype: int64